File: order_book.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto
from typing import Dict, List, Optional, Tuple
import heapq
import itertools
import time


class Side(Enum):
    """
    Represents the direction of an order.

    BUY  - the trader wants to purchase, matching against the best ask.
    SELL - the trader wants to sell, matching against the best bid.
    """
    BUY = auto()
    SELL = auto()

    def opposite(self) -> "Side":
        return Side.BUY if self is Side.SELL else Side.SELL


@dataclass
class Order:
    """
    Represents an order submitted to the limit order book.

    Attributes:
        order_id : Unique key for order.
        side     : BUY or SELL.
        price    : Limit price.
                   - For limit orders: a float (e.g., 101.5).
                   - For market orders: None = “execute immediately
                     at the best available price”.
        quantity : Number of units to buy or sell.
        timestamp: Used for FIFO priority when price is equal
                   (price-time priority).
    """
    order_id: int
    side: Side
    price: Optional[float]
    quantity: int
    timestamp: float


@dataclass
class Trade:
    """
    Represents an executed trade between a buy order and a sell order.

    Attributes:
        buy_order_id  : ID of the order on the buy side.
        sell_order_id : ID of the order on the sell side.
        price         : Execution price (resting order's price).
        quantity      : Amount traded.
        timestamp     : Time of execution.
    """
    buy_order_id: int
    sell_order_id: int
    price: float
    quantity: int
    timestamp: float


@dataclass
class OrderNode:
    """
    A node in the doubly linked list for a price level.

    Attributes:
        order : The Order object associated with this node.
        prev  : Pointer to the previous order node.
        next  : Pointer to the next node order node.
    """

    order: Order
    prev: Optional["OrderNode"] = None
    next: Optional["OrderNode"] = None


@dataclass
class PriceLevel:
    """
    Doubly linked list structure of resting orders at a single price.

    Attributes:
        head : First order node in the price level.
        tail : Last order node in the price level.
    """
    head: Optional[OrderNode] = None
    tail: Optional[OrderNode] = None


class OrderBook:
    """
    Simple single-instrument (one asset) limit order book with price-time priority:
    best price first, and FIFO within each price level.

    Separated data structures for price and orders:
    - Heaps store ONLY price levels (quick to find best price, smaller heaps).
    - Doubly linked lists (per price) store individual orders (for time priority).

    - Bids: max-heap on price (using min-heap of negative prices)
    - Asks: min-heap on price
    """
    def __init__(self) -> None:
        # PriceLevel doubly linked list (FIFO at each price)
        self.bids: Dict[float, PriceLevel] = {}
        self.asks: Dict[float, PriceLevel] = {}

        # Heaps of prices
        self.bid_heap: List[float] = []
        self.ask_heap: List[float] = []

        # Map order_id -> (side, price, OrderNode)
        self.order_map: Dict[int, Tuple[Side, float, OrderNode]] = {}

        # Trade log
        self.trades: List[Trade] = []

        # For incrementing order IDs
        self._order_id_counter = itertools.count(1)

    # ---------- Public API ----------

    def add_limit_order(
        self,
        side: Side,
        price: float,
        quantity: int,
        order_id: Optional[int] = None,
        timestamp: Optional[float] = None,
    ) -> int:
        """
        Add a limit order and immediately match against the opposite side.
        Returns the order ID (generated if not supplied).
        """
        if order_id is None:
            order_id = self._next_free_id()
        else:
            if order_id in self.order_map:
                raise ValueError(f"Order ID {order_id} already exists in the order book.")
        if timestamp is None:
            timestamp = time.time()

        order = Order(order_id, side, price, quantity, timestamp)
        self._match(order)

        # Rest remaining quantity after match on the book
        if order.quantity > 0:
            self._add_resting_order(order)

        return order_id

    def add_market_order(
        self,
        side: Side,
        quantity: int,
        order_id: Optional[int] = None,
        timestamp: Optional[float] = None,
    ) -> int:
        """
        Add a market order which matches against the best available prices
        until it is fully filled or book is empty. Any remaining quantity is discarded.
        """
        if order_id is None:
            order_id = next(self._order_id_counter)
        if timestamp is None:
            timestamp = time.time()

        order = Order(order_id, side, price=None, quantity=quantity, timestamp=timestamp)
        self._match(order)

        # No resting, ignore any remaining quantities
        return order_id

    def cancel_order(self, order_id: int) -> bool:
        """
        Cancel an existing resting order. Returns True if cancelled, False if not found
        or already fully executed.

        O(1) removal by unlinking node pointers in a doubly linked list.
        """
        info = self.order_map.get(order_id)
        if info is None:
            return False

        side, price, node = info
        book_side = self._book(side)
        level = book_side.get(price)
        if level is None:
            return False

        # Unlink node
        prev_node = node.prev
        next_node = node.next

        if prev_node:
            prev_node.next = next_node
        else:
            # Order node was head
            level.head = next_node

        if next_node:
            next_node.prev = prev_node
        else:
            # Order node was tail
            level.tail = prev_node

        # Remove if empty price level
        if level.head is None:
            del book_side[price]

        # Remove from order map
        del self.order_map[order_id]
        return True

    def best_bid(self) -> Optional[Tuple[float, int]]:
        """
        Return (price, total_quantity_at_price) for best bid, or None if no bids.
        """
        price = self._peek_best_price(Side.BUY)
        if price is None:
            return None
        book_side = self._book(Side.BUY)
        level = book_side[price]
        qty = self._sum_level_quantity(level)
        return price, qty

    def best_ask(self) -> Optional[Tuple[float, int]]:
        """
        Return (price, total_quantity_at_price) for best ask, or None if no asks.
        """
        price = self._peek_best_price(Side.SELL)
        if price is None:
            return None
        book_side = self._book(Side.SELL)
        level = book_side[price]
        qty = self._sum_level_quantity(level)
        return price, qty

    def get_depth(self, side: Side, levels: int = 5) -> List[Tuple[float, int]]:
        """
        Return top N price levels on given side as list of (price, total_quantity),
        sorted best-to-worst (descending for bids, ascending for asks).
        """
        book_side = self._book(side)

        if side == Side.BUY:
            prices = sorted(book_side.keys(), reverse=True)
        else:
            prices = sorted(book_side.keys())

        result: List[Tuple[float, int]] = []
        for p in prices[:levels]:
            level = book_side[p]
            qty = self._sum_level_quantity(level)
            result.append((p, qty))
        return result

    def get_trades(self) -> List[Trade]:
        """
        Return list of all trades executed so far.
        """
        return list(self.trades)

    # ---------- Internal helpers ----------

    def _book(self, side: Side) -> Dict[float, PriceLevel]:
        """
        Return the price -> PriceLevel mapping for the given side.
        """
        return self.bids if side == Side.BUY else self.asks

    def _heap(self, side: Side) -> List[float]:
        """
        Return the heap of prices for the given side.
        """
        return self.bid_heap if side == Side.BUY else self.ask_heap

    def _sum_level_quantity(self, level: PriceLevel) -> int:
        """
        Sum total quantity at a given price level by traversing its linked list.
        """
        total = 0
        node = level.head
        while node is not None:
            total += node.order.quantity
            node = node.next
        return total

    def _next_free_id(self) -> int:
        """
        Return the next available (non-duplicated) order ID from iterator.
        """
        while (oid := next(self._order_id_counter)) in self.order_map:
            pass
        return oid

    def _match(self, incoming: Order) -> None:
        """
        Match incoming order against the opposite side until it is fully filled,
        book is empty, or price limits are violated.
        """
        if incoming.side == Side.BUY:
            price_cmp = lambda best_price: (
                incoming.price is None or best_price <= incoming.price
            )
        else:
            price_cmp = lambda best_price: (
                incoming.price is None or best_price >= incoming.price
            )

        opposite_side = incoming.side.opposite()
        opposite_book = self._book(opposite_side)

        while incoming.quantity > 0:
            # Sweep through the best available price levels on the opposite side,
            # matching as much of the incoming order as possible, and
            # stop when incoming order is fully filled or no acceptable prices.
            best_price = self._pop_best_price_from_heap(opposite_side)
            if best_price is None:
                # No liquidity
                break

            if not price_cmp(best_price):
                # Push back the popped price if no acceptable price
                self._push_price(best_price, opposite_side)
                break

            level = opposite_book.get(best_price)
            if not level or level.head is None:
                opposite_book.pop(best_price, None)
                continue

            # Match first resting order at this price level (FIFO)
            while level.head is not None and incoming.quantity > 0:
                node = level.head
                resting = node.order

                traded_qty = min(incoming.quantity, resting.quantity)
                trade_price = best_price  # trade at resting order price
                self._record_trade(incoming, resting, trade_price, traded_qty)

                incoming.quantity -= traded_qty
                resting.quantity -= traded_qty

                if resting.quantity == 0:
                    # Remove this resting order from the head of the linked list
                    next_node = node.next
                    level.head = next_node
                    if next_node is not None:
                        next_node.prev = None
                    else:
                        # List became empty
                        level.tail = None
                    # Remove from order map
                    self.order_map.pop(resting.order_id, None)

            if level.head is None:
                # Remove if empty price level
                del opposite_book[best_price]
            else:
                # Still has liquidity, push back price into heap
                self._push_price(best_price, opposite_side)

    def _add_resting_order(self, order: Order) -> None:
        """
        Insert a partially filled or new order into its side of the book.
        """
        if order.price is None:
            raise ValueError("Resting market orders are not supported")

        book_side = self._book(order.side)

        # Get or create price level
        level = book_side.get(order.price)
        if level is None:
            level = PriceLevel()
            book_side[order.price] = level
            self._push_price(order.price, order.side)

        # Append to tail (FIFO)
        node = OrderNode(order=order)
        if level.tail is None:
            # Empty list
            level.head = level.tail = node
        else:
            level.tail.next = node
            node.prev = level.tail
            level.tail = node

        # Store node pointer in order map
        self.order_map[order.order_id] = (order.side, order.price, node)

    def _push_price(self, price: float, side: Side) -> None:
        """
        Insert a price into the appropriate heap.
        """
        if side == Side.BUY:
            heapq.heappush(self.bid_heap, -price)
        else:
            heapq.heappush(self.ask_heap, price)

    def _peek_best_price(self, side: Side) -> Optional[float]:
        """
        Peek best price without removing it.
        """
        heap = self._heap(side)
        book_side = self._book(side)

        while heap:
            raw = heap[0]
            price = -raw if side == Side.BUY else raw
            level = book_side.get(price)
            if level is not None and level.head is not None:
                return price
            else:
                # Lazy cleanup of stale price levels, ensures that trades are done
                # quicker, as heap removal of an arbitrary element is O(n).
                heapq.heappop(heap)
        return None

    def _pop_best_price_from_heap(self, side: Side) -> Optional[float]:
        """
        Pop and return best price, ensuring it still has liquidity.
        """
        heap = self._heap(side)
        book_side = self._book(side)

        while heap:
            raw = heapq.heappop(heap)
            price = -raw if side == Side.BUY else raw
            level = book_side.get(price)
            if level is not None and level.head is not None:
                return price
        return None

    def _record_trade(self, incoming: Order, resting: Order, price: float, qty: int) -> None:
        """
        Record an executed trade.
        """
        ts = time.time()
        if incoming.side == Side.BUY:
            buy_id, sell_id = incoming.order_id, resting.order_id
        else:
            buy_id, sell_id = resting.order_id, incoming.order_id

        self.trades.append(
            Trade(
                buy_order_id=buy_id,
                sell_order_id=sell_id,
                price=price,
                quantity=qty,
                timestamp=ts,
            )
        )

File: cpp/order_book.cpp
#include "order_book.hpp"

#include <algorithm>
#include <chrono>
#include <functional>
#include <stdexcept>


namespace lob {


using Clock = std::chrono::steady_clock;

// Return a monotonic timestamp in nanoseconds.
// Used for order and trade timestamps.
static std::int64_t now_ts() {
    auto now = Clock::now().time_since_epoch();
    return std::chrono::duration_cast<std::chrono::nanoseconds>(now).count();
}


// ============================================================================
// Public API (implementation only — documentation in .hpp)
// ============================================================================


OrderBook::OrderId OrderBook::addLimitOrder(
    Side side,
    Price price,
    Quantity quantity,
    std::optional<OrderId> order_id,
    std::optional<std::int64_t> ts_ns)
{
    OrderId id;
    if (!order_id) {
        id = nextFreeId();
    } else {
        id = *order_id;
        if (order_map_.find(id) != order_map_.end()) {
            throw std::invalid_argument(
                "Order ID" + std::to_string(id)
                + "already exists in the order book"
            );
        }
    }
    std::int64_t ts = ts_ns.value_or(now_ts());

    Order order{ id, side, price, quantity, ts };
    matchIncoming(order);

    if (order.quantity > 0) {
        addRestingOrder(order);
    }

    return id;
}


OrderBook::OrderId OrderBook::addMarketOrder(
    Side side,
    Quantity quantity,
    std::optional<OrderId> order_id,
    std::optional<std::int64_t> ts_ns)
{

    OrderId id;
    if (!order_id) {
        id = nextFreeId();
    } else {
        id = *order_id;
        if (order_map_.find(id) != order_map_.end()) {
            throw std::invalid_argument(
                "Order ID" + std::to_string(id)
                + "already exists in the order book"
            );
        }
    }
    std::int64_t ts = ts_ns.value_or(now_ts());

    Order order{ id, side, std::nullopt, quantity, ts };
    matchIncoming(order);

    return id;
}


bool OrderBook::cancelOrder(OrderId id) {
    auto it = order_map_.find(id);
    if (it == order_map_.end()) {
        return false;
    }

    const OrderInfo& info = it->second;
    Side side       = info.side;
    double price    = info.price;
    OrderNode* node = info.node;

    PriceMap& book_side = book(side);

    auto level_it = book_side.find(price);
    if (level_it == book_side.end()) {
        return false;
    }

    PriceLevel& level = level_it->second;

    // optimise later with cpp's built in doubly linked list - std::list
    OrderNode* prev = node->prev;
    OrderNode* next = node->next;

    if (prev) prev->next = next;
    else      level.head = next;

    if (next) next->prev = prev;
    else      level.tail = prev;
    // --

    if (!level.head) {
        book_side.erase(level_it);
    }

    order_map_.erase(it);
    delete node;
    return true;
}


std::optional<std::pair<OrderBook::Price, OrderBook::Quantity>>
OrderBook::bestBid() const
{
    auto price_opt = peekBestPrice(Side::Buy);
    if (!price_opt) return std::nullopt;

    Price price = *price_opt;
    const auto& book_side = book(Side::Buy);

    const PriceLevel &level = book_side.at(price);

    return { { price, sumLevelQuantity(level) } };
}

std::optional<std::pair<OrderBook::Price, OrderBook::Quantity>>
OrderBook::bestAsk() const
{
    auto price_opt = peekBestPrice(Side::Sell);
    if (!price_opt) return std::nullopt;

    Price price = *price_opt;
    const auto& book_side = book(Side::Sell);

    const PriceLevel &level = book_side.at(price);

    return { { price, sumLevelQuantity(level) } };
}


std::vector<std::pair<OrderBook::Price, OrderBook::Quantity>>
OrderBook::getDepth(Side side, std::size_t levels) const
{
    std::vector<std::pair<Price, Quantity>> result;

    const PriceMap& book_side = book(side);
    if (book_side.empty() || levels == 0) {
        return result;
    }

    std::vector<Price> prices;
    prices.reserve(book_side.size());
    for (const auto& kv : book_side) {
        prices.push_back(kv.first);
    }

    if (side == Side::Buy) {
        std::sort(prices.begin(), prices.end(), std::greater<Price>());
    } else {
        std::sort(prices.begin(), prices.end());
    }

    std::size_t count = std::min(levels, prices.size());
    result.reserve(count);

    for (std::size_t i = 0; i < count; ++i) {
        Price p = prices[i];
        const PriceLevel& level = book_side.at(p);
        result.emplace_back(p, sumLevelQuantity(level));
    }

    return result;
}


void OrderBook::clear() {
    // Free all OrderNode memory on both sides
    auto free_side = [](PriceMap& m) {
        for (auto& [price, level] : m) {
            OrderNode* node = level.head;
            while (node) {
                OrderNode* next = node->next;
                delete node;
                node = next;
            }
        }
        m.clear();
    };

    free_side(bids_);
    free_side(asks_);

    // Reset heaps by reassigning empty ones
    bid_heap_ = PriceQueue{};
    ask_heap_ = MinPriceQueue{};

    order_map_.clear();
    trades_.clear();
    next_order_id_ = 1;
}


// ============================================================================
// Internal Helpers
// ============================================================================


// Simple side selector
OrderBook::PriceMap& OrderBook::book(Side side) {
    return (side == Side::Buy) ? bids_ : asks_;
}


const OrderBook::PriceMap& OrderBook::book(Side side) const {
    return (side == Side::Buy) ? bids_ : asks_;
}


OrderBook::Quantity OrderBook::sumLevelQuantity(const PriceLevel& level) {
    Quantity total = 0;
    for (OrderNode* node = level.head; node; node = node->next) {
        total += node->order.quantity;
    }
    return total;
}


OrderBook::OrderId OrderBook::nextFreeId() {
    // Increment until an unused ID is found.
    while (order_map_.find(next_order_id_) != order_map_.end()) {
        ++next_order_id_;
    }
    return next_order_id_++;
}


void OrderBook::matchIncoming(Order& incoming) {
    // Check whether the incoming order crosses the best price.
    auto price_ok = [&](Price best_price) {
        if (!incoming.price) return true; // market order
        double limit = *incoming.price;
        return (incoming.side == Side::Buy) ? (best_price <= limit)
                                            : (best_price >= limit);
    };

    Side opp = opposite(incoming.side);
    PriceMap& opp_book = book(opp);

    while (incoming.quantity > 0) {
        auto best_price_opt = popBestPrice(opp);
        if (!best_price_opt) break;

        Price best_price = *best_price_opt;

        // No acceptable price
        if (!price_ok(best_price)) {
            pushPrice(best_price, opp);
            break;
        }

        auto lvl_it = opp_book.find(best_price);
        if (lvl_it == opp_book.end() || !lvl_it->second.head) {
            continue;
        }

        PriceLevel& level = lvl_it->second;

        // FIFO matching
        while (level.head && incoming.quantity > 0) {
            OrderNode* node = level.head;
            Order& resting = node->order;

            Quantity qty = std::min(incoming.quantity, resting.quantity);
            recordTrade(incoming, resting, best_price, qty);

            incoming.quantity -= qty;
            resting.quantity  -= qty;

            // Remove fully filled resting order
            if (resting.quantity == 0) {
                OrderNode* next = node->next;
                level.head = next;
                if (next) next->prev = nullptr;
                else      level.tail = nullptr;

                order_map_.erase(resting.order_id);
                delete node;
            }
        }

        // Remove empty price level
        if (!level.head) {
            opp_book.erase(best_price);
        } else {
            // Still liquidity; put price level back into heap
            pushPrice(best_price, opp);
        }
    }
}


void OrderBook::addRestingOrder(const Order& order) {
    if (!order.price) {
        throw std::runtime_error("Resting market orders are not supported");
    }

    Price p = *order.price;
    PriceMap& book_side = book(order.side);

    // Get or create new price level
    auto [it, created] = book_side.try_emplace(p, PriceLevel{});
    PriceLevel& level = it->second;

    if (created) {
        pushPrice(p, order.side);
    }

    // FIFO
    OrderNode* node = new OrderNode{order, nullptr, nullptr};

    if (!level.tail) {
        level.head = level.tail = node;
    } else {
        level.tail->next = node;
        node->prev = level.tail;
        level.tail = node;
    }

    order_map_[order.order_id] = OrderInfo{order.side, p, node};
}


void OrderBook::pushPrice(Price price, Side side) {
    // No reverse vs Python (heaps are already max and min)
    if (side == Side::Buy) bid_heap_.push(price);
    else                   ask_heap_.push(price);
}


std::optional<OrderBook::Price> OrderBook::peekBestPrice(Side side) const {
    auto* self = const_cast<OrderBook*>(this);
    PriceMap& book_side = self->book(side);

    if (side == Side::Buy) {
        while (!self->bid_heap_.empty()) {
            Price p = self->bid_heap_.top();

            auto it = book_side.find(p);
            if (it != book_side.end() && it->second.head != nullptr) {
                return p;
            }

            self->bid_heap_.pop();
        }
    } else {
        while (!self->ask_heap_.empty()) {
            Price p = self->ask_heap_.top();

            auto it = book_side.find(p);
            if (it != book_side.end() && it->second.head != nullptr) {
                return p;
            }

            self->ask_heap_.pop();
        }
    }

    return std::nullopt;
}


std::optional<OrderBook::Price> OrderBook::popBestPrice(Side side) {
    PriceMap& book_side = book(side);

    if (side == Side::Buy) {
        while (!bid_heap_.empty()) {
            Price p = bid_heap_.top();
            bid_heap_.pop();

            auto it = book_side.find(p);
            if (it != book_side.end() && it->second.head != nullptr) {
                return p;
            }
        }
    } else {
        while (!ask_heap_.empty()) {
            Price p = ask_heap_.top();
            ask_heap_.pop();

            auto it = book_side.find(p);
            if (it != book_side.end() && it->second.head != nullptr) {
                return p;
            }
        }
    }

    return std::nullopt;
}


void OrderBook::recordTrade(const Order& incoming,
                            const Order& resting,
                            Price price,
                            Quantity qty)
{
    // Determine correct buy/sell IDs based on trade direction
    std::int64_t buy_id =
        (incoming.side == Side::Buy) ? incoming.order_id : resting.order_id;

    std::int64_t sell_id =
        (incoming.side == Side::Sell) ? incoming.order_id : resting.order_id;

    trades_.push_back(Trade{
        buy_id,
        sell_id,
        price,
        qty,
        now_ts()
    });
}


} // namespace lob
